### DISCUSSION TITLE
Confusion about instructions

### DISCUSSION THREAD

**Discussion by Jano-1** [ryanoasis/nerd-fonts#847] [Q&A] [answered]
I wanted to patch my own font, so from the instructions on the main page, I installed fontforge. Then going down the page a bit, it says:

> Alternative usage: Execute the patcher with the FontForge binary using the script flag:
> 
> fontforge -script font-patcher PATH_TO_FONT
> 

So I tried _fontforge -script font-patcher PATH_TO_FONT_ because none of the other instructions were applicable to me. It didn't work. Looking around I saw someone who had the same issue and it turned out they didn't use an argument. So I added _--complete_. It didn't work. I went down to examples: 

> ./font-patcher Droid\ Sans\ Mono\ for\ Powerline.otf
> ./font-patcher Droid\ Sans\ Mono\ for\ Powerline.otf -s -q
> ./font-patcher Droid\ Sans\ Mono\ for\ Powerline.otf --use-single-width-glyphs --quiet
> ./font-patcher Droid\ Sans\ Mono\ for\ Powerline.otf -w
> ./font-patcher Droid\ Sans\ Mono\ for\ Powerline.otf --windows --quiet
> ./font-patcher Droid\ Sans\ Mono\ for\ Powerline.otf --windows --pomicons --quiet
> 
> ./font-patcher Inconsolata.otf --fontawesome
> ./font-patcher Inconsolata.otf --fontawesome --octicons --pomicons
> ./font-patcher Inconsolata.otf
> 
> ./FontForge.AppImage -script /tmp/nerdfonts/font-patcher /tmp/nerdfonts/CascadiaMonoPL-Semibold.ttf --fontawesome -out /tmp
> ./FontForge.AppImage -script $PWD/font-patcher $PWD/CascadiaMonoPL-Semibold.ttf --octicons -out $HOME
> 
> docker run --rm -v ~/myfont/patchme:/in -v ~/myfont/patched:/out nerdfonts/patcher
> docker run --rm -v ~/Desktop/myfont/patchme:/in -v ~/Desktop/myfont/patched:/out nerdfonts/patcher --fontawesome

None of these applied to me. Googling I finally found this page: https://github.com/alexandermckay/font-patcher
Following the instructions on that page I realized I had to have the _src_ directory, and also the font-patcher file.

Am I missing something? Am I doing something wrong? What I did isn't included anywhere on this repo. I'm confused.

**Comment by Finii** [comment, marked as answer]
Sorry that the documentation is that misleading :-(

Lets first solve your problem, then tackle correction of the documentation.

The Nerd Fonts `font-patcher` is a script that tells `fontforge` how to patch your font. In order to add some symbols to your font it also needs that symbols - these binary files are not included in the `font-patcher` script. 

### Preparation for self-patching

For most (all except the Container method) you need [`fontforge`](https://fontforge.org). How you can get this depends on your OS. If you use some Linux, most probably some version of `fontforge` is in your distributions package repository and can be installed by your normal software installation way. Otherwise, in my opinion it is best to follow their (`fontforge`'s) installation instructions.
There are bugs in older versions - if you want to patch a very 'feature rich' font the patching might fail. Make sure to use at least the `fontforge` _March 2022 Release_ in those cases.

#### Recommended way
Currently the recommended way to get the `font-patcher` script is the Release page. There is a special zip-archive that for that:
https://github.com/ryanoasis/nerd-fonts/releases/download/2.2.0-RC/FontPatcher.zip

That zip file contains the script and the symbols. It has to be extracted to some directory and will put the symbol files (in form of symbol source fonts) in a sub-directory. The relative path between the script location and those files is hard-coded in the script (although it can be overridden for advanced usage).

#### Direct git access way
A good option, and needed if you want to patch with the most recent bugfixes applied (i.e. with HEAD), you need to download the `font-patcher` script directly from [here](https://github.com/ryanoasis/nerd-fonts/blob/master/font-patcher) and additionally all the content below `src/glyphs/` (from [here](https://github.com/ryanoasis/nerd-fonts/tree/master/src/glyphs)). It is a little bit hard(er) to download that many files manually, so this most likely is only a solution for you if you know how to work with git.
That files must be put in the prescribed relative-to-the-script sub-directory (i.e. `src/glyphs/`) or you have to specify the glyph directory in the `font-patcher` call.

#### Future simple way
There is a PR https://github.com/ryanoasis/nerd-fonts/pull/741 that, if you start `font-patcher` and it can not find all the source glyphs, downloads them automatically for you. This is the most convenient maybe for users that just download the script from the front page.

#### Container way
There is a docker container prepared [here](https://hub.docker.com/r/nerdfonts/patcher). You need to have docker installed on you system and then you can patch your font in that container. (I never used that.) It contains (I believe) `fontforge` and the source symbols, so that is a one stop solution.

### Actual patching
With everything downloaded and installed, the simplest way is on the command line to go in the same directory where `font-patcher`. I usually put the font-to-be-patched into the same directory. In principle `font-patcher` can be called on its own (and it will then call `fontforge` to do the actual patching) but for me this is not working very reliably on all platforms. A more robust way is to call `fontforge` from the command line and specify `font-patcher` as option (and `fontforge` will execute the script then).

```
cd wherever-font-patcher-is
./font-patcher --complete myfont.otf

[ ... or: ]

fontforge --script font-patcher --complete myfont.otf
```

You can give different options to the call, the most used are possibly `--complete` (patch in everything we have), and depending on your terminal emulation `--mono`. On MS Windows you might want to also specify `--windows`. The [other options](https://github.com/ryanoasis/nerd-fonts#font-patcher) are seldom needed.

The patched font will be created in you current workdir.

The container solution is similar but slightly different, see the manual on [dockerhub](https://hub.docker.com/r/nerdfonts/patcher).

----

For your concrete problems, what you do not say is how you got `font-patcher`? And maybe you can tell us the messages you got. Most probably you just downloaded the script file (`font-patcher`) and (or course!) not the `src/glyphs/` directory.

**Reply by Jano-1** [reply]
Thanks a lot for your detailed instructions, they are much appreciated. I got everything to work and patched the fonts successfully. Thank you!
